### PR TITLE
#2139 gantt chart package

### LINKED ISSUE #2139: Custom Gantt Chart Package
### Spike Type

Technical

### Goal

The goal of this spike is to make a custom gantt chart component that we can use for all our gantt chart needs

### Reason for Spike

The current gantt package sucks, so we're gonna make it custom

### PR DESCRIPTION
**Pull request by RChandler234**
## Changes

I've got the dependencies playing nice and the code ported over, but none of the styling is being applied

## Screenshots
Original (Parth's Implementation)
image
image

Mondays Only
image
image

3/26 Updates
image
image
image

4/10 Updates
image
image

## Checklist

It can be helpful to check the `Checks` and `Files changed` tabs.
Please review the [contributor guide](https://nerdocs.atlassian.net/wiki/spaces/NER/pages/8060929/Software+Contributor+Guide) and reach out to your Tech Lead if anything is unclear.
Please request reviewers and ping on slack only after you've gone through this whole checklist.

- [ ] All commits are tagged with the ticket number
- [ ] No linting errors / newline at end of file warnings
- [ ] All code follows repository-configured prettier formatting
- [ ] No merge conflicts
- [ ] All checks passing
- [ ] Screenshots of UI changes (see Screenshots section)
- [ ] Remove any non-applicable sections of this template
- [ ] Assign the PR to yourself
- [ ] No `yarn.lock` changes (unless dependencies have changed)
- [ ] Request reviewers & ping on Slack
- [ ] PR is linked to the ticket (fill in the closes line below)

Closes #2139  (issue #)

### CODE REVIEW

**Review by RChandler234 — COMMENTED** [review]
Can you also delete the old Gantt Chart Package code while you're in here and then move all this Gantt Chart code into the GanttPage folder

**Comment by RChandler234** [comment]
Love the filter thing you did, should make a mobile view for this super easy as well. Colors on the actual chart should match the colors you're using here (I like the legend colors, but use whichever looks better on the graph) Getting this weird scrollbar on it rn tho: 
image
Can you hide this scrollbar (one on the bottom):
image
Not sure if this is the place to address this, but don't love our bolded font choice, imo is kind of hard to read, might be worth changing:
image
Try rounding the outline of a team the same amount that project/wp bars are rounded, also try throwing some drop shadows on the edit outline to make it pop more (also see how it looks or the project/wp bars, might cool, but could also see it being too much):
image
Also, I don't think the team title should move when u scroll (if that feels worse to use lmk tho)

**Review by RChandler234 — COMMENTED** [review]
this should be sticky when u scroll down
image

**Comment by RChandler234** [comment]
filters button looks weird next to all the legend stuff now, maybe we make it a filter icon button instead?
image

**Comment by RChandler234** [comment]
There's a decent bit of empty space between the top row and the line of dates, wondering if we could do anything to close that gap (by moving the year info somewhere). Also the search bar should p be a bit thinner/ aligned better
image

**Review by RChandler234 — COMMENTED** [review]
need to make sure light mode looks decent as well
image

**Comment by RChandler234** [comment]
this is happening when you scroll to the left, then scroll down:
image

**Comment by RChandler234** [comment]
only the dates should be sticky, not the title/ search bar
image

File: src/frontend/src/pages/GanttPage/Gantt.tsx [hunk 1]
```diff
@@ -0,0 +1,229 @@
...
+  );
+}
+
+function Event({
+  days,
+  event,
+  createChange,
+  isEditMode,
+  ...props
+}: { days: Date[]; event: Task; createChange: (change: EventChange) => void; isEditMode: boolean } & ComponentProps<'div'>) {
```
Comment (RChandler234) [inline src/frontend/src/pages/GanttPage/Gantt.tsx:60]: ok next steps are naming and organizing this code a lot better, so people can do work in it. It looks like this file is for one team section on the Gantt Chart (correct me if I'm wrong). This event component should be moved into another file and called GanttChartElement or GanttChartActivity or something (it should also be changed to be an arrow method

File: src/frontend/src/pages/GanttPage/Gantt.tsx [hunk 2]
```diff
@@ -0,0 +1,229 @@
...
+
+interface GanttProps {
+  start: Date;
+  end: Date;
+  tasks: Task[];
+  isEditMode: boolean;
+  saveChanges: (eventChanges: EventChange[]) => void;
+}
+
+export function Gantt({ start, end, tasks, isEditMode, saveChanges }: GanttProps) {
```
Comment (RChandler234) [inline src/frontend/src/pages/GanttPage/Gantt.tsx:24]: see comment below, but if this represents a team in the gantt chart, it should be named GanttChartSection or GanttChartTeamSection, be an arrow function component and be in it's own file

File: src/frontend/src/pages/GanttPage/GanttChart.tsx [hunk 3]
```diff
@@ -4,29 +4,21 @@
...
   ganttTasks: Task[];
+  start: Date;
+  end: Date;
   onExpanderClick: (ganttTasks: Task) => void;
+  isEditMode: boolean;
+  saveChanges: (eventChanges: EventChange[]) => void;
 }
 
-const GanttChart: React.FC<GanttPageProps> = ({ ganttTasks, onExpanderClick }) => {
+const GanttChart: React.FC<GanttPageProps> = ({ ganttTasks, start, end, onExpanderClick, isEditMode, saveChanges }) => {
```
Comment (RChandler234) [inline src/frontend/src/pages/GanttPage/GanttChart.tsx:19]: I think this file could be integrated into the GanttChartSection Component, see my other comment

File: src/frontend/src/pages/GanttPage/GanttChartCalendar.tsx [hunk 4]
```diff
@@ -0,0 +1,78 @@
+import { Box, Typography, Card, useTheme } from '@mui/material';
+import { eachDayOfInterval, isMonday, format, getDate } from 'date-fns';
+import { GANTT_CHART_GAP_SIZE, GANTT_CHART_CELL_SIZE } from '../../utils/gantt.utils';
+
+interface GanttChartCalendarProps {
+  start: Date;
+  end: Date;
+}
+
+export function GanttChartCalendar({ start, end }: GanttChartCalendarProps) {
```
Comment (RChandler234) [inline src/frontend/src/pages/GanttPage/GanttChartCalendar.tsx:10]: Let's rename this to GanttChartTimeline

File: src/frontend/src/pages/GanttPage/GanttPageFilter.tsx [hunk 5]
```diff
@@ -2,123 +2,89 @@
...
+        <Chip
+          label={buttonText}
+          sx={{ minWidth: '150px', borderRadius: '20px', backgroundColor: theme.palette.primary.main }}
+        />
+      }
+    />
+  );
+};
+
 const GanttPageFilter: FC<GanttPageFilterProps> = ({
```
Comment (RChandler234) [inline src/frontend/src/pages/GanttPage/GanttPageFilter.tsx:53]: Rename to GanttChartFilters

File: src/frontend/src/pages/GanttPage/GanttPageWrapper.tsx [hunk 6]
```diff
@@ -20,50 +20,74 @@ import {
...
+  GanttWorkPackageTextColorPipe,
+  WbsElementStatusTextPipe,
+  WorkPackageStageTextPipe
+} from '../../utils/enum-pipes';
+import { SearchBar } from '../../components/SearchBar';
 
 /**
  * Documentation for the Gantt package: https://github.com/MaTeMaTuK/gantt-task-react
  */
 const GanttPageWrapper: FC = () => {
```
Comment (RChandler234) [inline src/frontend/src/pages/GanttPage/GanttPageWrapper.tsx:41]: This component is a beast, I would if possible break it out into a GanttChartPage component that wraps a component named GanttChart. Try keeping a lot of the logic/ query param stuff in GanttChartPage, and a lot of the React, component stuff in GanttChart. You could also probably break some of these components out into their own files

Comment (RChandler234) [reply src/frontend/src/pages/GanttPage/GanttPageWrapper.tsx:41]: I would make a folder called GanttChartComponents, under the GanttChartPage folder and put any components (e.g. the GanttChartElement component or GanttChartFilters) that don't contain any higher level organizational logic and are pretty isolated in their functionality/ placement on the screen.

**Review by walker-sean — DISMISSED** [review]

**Pull request merged**